File: the-c-programming-language/CP5/qsort.js
// js version of quick sort.

const swap = (arr, i, j) => {
  const temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

const qsort = (arr, left, right) => {
  if (left >= right)
    return

  let i
  let last
  swap(arr, left, Math.floor((left + right) / 2))
  last = left
  for (i = left + 1; i <= right; i++) {
    if (arr[i] < arr[left])
      swap(arr, i, ++last)
  }
  swap(arr, left, last)
  qsort(arr, left, last - 1)
  qsort(arr, last + 1, right)
}

const arr = [4, 5, 3, 1, 8, 6, 4, 8, 9, 2, 6]

qsort(arr, 0, arr.length - 1)

console.log(arr)

File: the-c-programming-language/CP5/qsort.c
#include <string.h>
#include <stdio.h>

void qsort(char *v[], int left, int right);
void swap(char *v[], int i, int j);

void qsort(char *v[], int left, int right)
{
  int i, last;
  if (left >= right)
    return;
  swap(v, left, (left + right) / 2);
  last = left;
  for (i = left + 1; i <= right; i++) {
    if (strcmp(v[i], v[left]) < 0) {
      swap(v, ++last, i);
    }
  }
  swap(v, left, last);
  qsort(v, left, last - 1);
  qsort(v, last + 1, right);
}

void swap(char *v[], int i, int j)
{
  char *temp;
  temp = v[i];
  v[i] = v[j];
  v[j] = temp;
}

int main(void)
{
  char *v[] = {
    "hello",
    "nice",
    "wujingwen",
    "jocs",
    "ransixi"
  };
  int i;

  qsort(v, 0, 4);
  for (i = 0; i < 5; i++) {
    printf("%s\n", v[i]);
  }
  return 0;
}